go-controller/pkg/libovsdbops/db_object_ids.go

```diff
@@ -0,0 +1,283 @@
+}
+
+func (objectIDs *DbObjectIDs) NewDbObjectWithoutIDs(idsToDelete ...ExternalIDKey) *DbObjectIDs {
+	ids := deepcopyMap(objectIDs.objectIDs)
+	for _, keyToDel := range idsToDelete {
+		delete(ids, keyToDel)
+	}
+	return NewDbObjectIDs(objectIDs.idsType, objectIDs.ownerControllerName, ids)
+}
+
+func (objectIDs *DbObjectIDs) HasSameOwner(ownerController string, indexT ObjectIDsType) bool {
+	return objectIDs.ownerControllerName == ownerController && objectIDs.idsType.IsSameType(indexT)
+}
+
+// GetObjectID returns value from objectIDs.objectIDs map, and empty string for not found values.
+// Usually objectIDs.objectIDs doesn't include PrimaryIDKey, OwnerTypeKey, and OwnerControllerKey.
+func (objectIDs *DbObjectIDs) GetObjectID(key ExternalIDKey) string {
+	return objectIDs.objectIDs[key]
+}
+
+// GetExternalIDs returns a map of ids, that always includes keys
+// - OwnerControllerKey
+// - OwnerTypeKey
+// - PrimaryIDKey (constructed by DbObjectIDs.GetUniqueID)
+// and also all keys that are used by objectIDs.idsType.
+// If some required keys are not present in the DbObjectIDs.objectIDs, they will be replaced with empty strings.
+// That may lead to having entries with the same primary id, so make sure to always set all required ids before
+// calling this method unless empty ids are intended.
+func (objectIDs *DbObjectIDs) GetExternalIDs() map[string]string {
+	externalIDs := objectIDs.getExternalIDsWithoutPrimary(false)
+	externalIDs[PrimaryIDKey.String()] = objectIDs.GetUniqueID()
+	return externalIDs
+}
+
+// GetUniqueID returns a unique string that identifies the object with given DbObjectIDs.
+// Different DbIndexes should always have different ids, equal DbIndexes will return the same string.
+// Returned string start with "<dbIndex.ownerControllerName>:<dbIndex.idsType.ownerObjectType>"
+// After that values from DbObjectIDs.objectIDs are added in order set in dbIDsMap, and are separated by ":".
+// If some required keys are not present in the DbObjectIDs.objectIDs, they will be replaced with empty strings.
+func (objectIDs *DbObjectIDs) GetUniqueID() string {
+	id, _ := objectIDs.getUniqueIDWithCheck()
+	return id
+}
+
+// getUniqueIDWithCheck returns primary id that is build based on present objectIDs values.
+// Some ids required by DbObjectIDs.idsType may be missing, in that case id still will be returned, but
+// missing values will be replaced with empty strings, therefore that id may not be unique.
+// If at least one required key was missing, an error will be returned.
+func (objectIDs *DbObjectIDs) getUniqueIDWithCheck() (string, error) {
+	id := objectIDs.ownerControllerName + ":" + string(objectIDs.idsType.ownerObjectType)
+	var err error
+	for _, key := range objectIDs.idsType.GetExternalIDKeys() {
+		value, ok := objectIDs.objectIDs[key]
+		if !ok {
+			err = fmt.Errorf("key %v is required but not present", key)
+		}
+		id += ":" + value
+	}
+	return id, err
```

Comment:
Why would we not return an empty id string on error?

Reply:
I guess I am still missing something here. 

In which cases are we building a usable unique ID without providing all the keys?

Reply:
yeah so the only use case that is left for this is `(objectIDs *DbObjectIDs) String()` method which is used by addresssetfactory just to generate a unique id based on a subset of ids.
I think it's ok to just move that starnge functionality to the `String()` function itself and forget about it.

Reply:
OOOK so it is merely for a string representation of DbObjectIDs that is not used as unique id. 

I guess I was really confused by the naming on that code re-usage pattern.

We could have `toString(bool allKeysOrError) (string, error)` that could be used for `String()` with false or `getUniqueID()' with true. Anyway its fine now.